### ISSUE TITLE
Error with using Pytorch

### ISSUE THREAD

**Issue by hoba87** [FitSNAP/FitSNAP#175]
I am using the latest version. With linear and quadratic fit fitsnap works, but when I try to use pytorch I get the following error. The shape size 96 seems to have something to do with the number of atoms in one configuration. The shape size 8 is changed if I change the number of groups.
My pytorch section looks like:

```
[PYTORCH]
layer_sizes = 60 60 1
learning_rate = 1.5e-4
num_epochs = 100
batch_size = 1
save_state_output = TiO2_pytorch_output.pt
energy_weight = 1.0
force_weight = 1.0
multi_element_option = 1
```
```
File "/opt/FitSNAP/fitsnap3lib/calculators/lammps_base.py", line 91, in process_configs_nonlinear
    self._collect_lammps_nonlinear()
  File "/opt/FitSNAP/fitsnap3lib/calculators/lammps_snap.py", line 177, in _collect_lammps_nonlinear
    self.pt.shared_arrays['a'].array[index:index+bik_rows] = bispectrum_components
ValueError: could not broadcast input array from shape (96,495) into shape (8,495)
```

**Comment by rohskopf** [comment]
@hoba87 it looks like fitsnap thinks there are 8 atoms in this configuration, but there are actually 96.

It might be an issue with how things are set up, or it could be a bug. I could quickly debug this if I can have the files (input script and group files) that cause it. Can you share a minimum reproducible example?

**Comment by hoba87** [comment]
Here you find a minimal example:

[example.zip](https://github.com/FitSNAP/FitSNAP/files/10767781/example.zip)

**Comment by rohskopf** [comment]
This error occurred because we haven't tried `quadraticflag = 1` with NNs before. I should raise an error in case someone else tries this.

A couple more changes needed in the input script:

```
[BISPECTRUM]
numTypes = 2  # number of atom types
type = Ti O   # atom type symbols, in order of lammps type number
twojmax = 6 6		# number of bispectrum coefficients
rcutfac = 6.0		# cut off / Angstrom
rfac0 = 0.99363
rmin0 = 0.0
wj = 0.7 1.0        # neighbour weights for each element to distinguish element types
radelem = 0.5 0.5   # additional cutoff factor for each element, effective cutoff is 2*rcutfac*radelem
quadraticflag = 0
bzeroflag = 1
bikflag = 1
dgradflag = 1

[CALCULATOR]
calculator = LAMMPSSNAP
energy = 1
per_atom_energy = 1
force = 1
stress = 0
nonlinear = 1  # 1 for pytorch

[ESHIFT]
Ti = 0.0
O = 0.0

[PYTORCH]
layer_sizes = num_desc 60 60 1
learning_rate = 1e-4 # good range to try: 1e-5 - 1e-4
num_epochs = 1000
batch_size = 4
save_state_output = TiO2_pytorch_output.pt
# save_state_input = TiO2_pytorch_input.pt
multi_element_option = 2

[SOLVER]
solver = PYTORCH      # SVD, PYTORCH
compute_testerrs = 1
detailed_errors = 1

[SCRAPER]
scraper = JSON

[PATH]
dataPath = data

[OUTFILE]
metrics = TiO2-metrics.md
potential = TiO2_snap

[REFERENCE]
units = metal
pair_style = zero 10.0
pair_coeff = * *

[GROUPS]
group_sections = name training_size testing_size eweight fweight vweight
group_types = str float float float float float
smartweights = 0
random_sampling = 0
job_1592_(TiO2)32 = 		1.0  0.0 	1e-2 1.0  1

[EXTRAS]
dump_peratom = 0 # for dumping detailed per-atom info (e.g. forces)
dump_perconfig = 0 # for dumping detailed per-config info (e.g. energies)

[MEMORY]
override = 0
```

This gave me the following errors (use most recent version for detailed errors with NNs):

```
Group              Train/Test   Property     Count       MAE         RMSE   
*ALL                 Train       Energy       19     1.127e-02   1.128e-02
*ALL                 Train       Force       1824    1.607e-01   2.064e-01
```

Energy MAE = 11.3 meV/atom
Force MAE = 160.7 meV/A

NOTE: There might be some confusion in the cutoff; the effective cutoff is `2*rcutfac*radelem`. I will update the docs on this, because it isn't clear currently.

**Comment by hoba87** [comment]
Ok thanks, that seems to work, at least for the training (GPU is used).
At the end however I get another error, related to a memory issue:

```
Traceback (most recent call last):
  File "/opt/FitSNAP/fitsnap3/__main__.py", line 48, in main
    snap.perform_fit()
  File "/opt/FitSNAP/fitsnap3lib/fitsnap.py", line 152, in perform_fit
    error_analysis()
  File "/opt/FitSNAP/fitsnap3lib/parallel_tools.py", line 277, in timed
    result = method(*args, **kw)
  File "/opt/FitSNAP/fitsnap3lib/fitsnap.py", line 148, in error_analysis
    self.solver.error_analysis()
  File "/opt/FitSNAP/fitsnap3lib/solvers/solver.py", line 338, in error_analysis
    decorated_error_analysis()
  File "/opt/FitSNAP/fitsnap3lib/parallel_tools.py", line 308, in check_if_rank_zero
    return method(*args, **kw)
  File "/opt/FitSNAP/fitsnap3lib/solvers/solver.py", line 159, in decorated_error_analysis
    (energies_model, forces_model) = self.evaluate_configs(option=1, standardize_bool=False, dtype=torch.float64)
  File "/opt/FitSNAP/fitsnap3lib/solvers/pytorch.py", line 711, in evaluate_configs
    (energies,forces) = decorated_evaluate_configs()
  File "/opt/FitSNAP/fitsnap3lib/parallel_tools.py", line 316, in check_if_rank_zero
    return method(*args, **kw)
  File "/opt/FitSNAP/fitsnap3lib/solvers/pytorch.py", line 674, in decorated_evaluate_configs
    self.device, dtype)
  File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/FitSNAP/fitsnap3lib/lib/neural_networks/pytorch.py", line 131, in forward
    per_atom_energies[elem_indices == i] = self.networks[elem](x[elem_indices == i]).flatten()
  File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/modules/container.py", line 119, in forward
    input = module(input)
  File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 94, in forward
    return F.linear(input, self.weight, self.bias)
  File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/functional.py", line 1753, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: Tensor for argument #2 'mat1' is on CPU, but expected it to be on GPU (while checking arguments for addmm)
```

**Comment by rohskopf** [comment]
Thanks @hoba87 , fixed in #176

**Comment by hoba87** [comment]
Ok now that is working, thank you for your support! I just noticed a huge memory peak at the end of the training, doubling the required memory, perhaps that is not absolutely necessary. However I will close the issue here.
